### PR TITLE
The fields speed, lat, lon and course are floats now

### LINKED ISSUE #54: Speed and other fields returning unexpected values
For some fields I'm getting speed and other field values as a integer rather than a float.

Looking at this segment https://github.com/M0r13n/pyais/blob/cc47fb1bf56eaabceb1657770d986a3172bac19a/pyais/messages.py#L794-L818

I'm somewhat confused as the type given to `bit_field` is `int` for everything while properties like speed/lat/lon etc should all be floats.
Also shouldn't the defaults be `None` for missing values, or am I misinterpreting what the default property represents? 
If it is meant to be 0 and not None the very least it should be 0.0 for floats and not 0

Please can you explain the use of `default` in the `bit_field` function.

### PR DESCRIPTION
**Pull request by M0r13n**
Closes #54

### CODE REVIEW

File: pyais/messages.py [hunk 1]
```diff
@@ -961,10 +964,10 @@ class MessageType22Broadcast(Payload):
     # If the message is broadcast (addressed field is 0),
     # the ne_lon, ne_lat, sw_lon, and sw_lat fields are the
     # corners of a rectangular jurisdiction area over which control parameter
-    ne_lon = bit_field(18, int, from_converter=from_course, to_converter=to_course, default=0, signed=True)
-    ne_lat = bit_field(17, int, from_converter=from_course, to_converter=to_course, default=0, signed=True)
-    sw_lon = bit_field(18, int, from_converter=from_course, to_converter=to_course, default=0, signed=True)
-    sw_lat = bit_field(17, int, from_converter=from_course, to_converter=to_course, default=0, signed=True)
+    ne_lon = bit_field(18, float, from_converter=from_course, to_converter=to_course, default=0, signed=True)
```
Comment (jeremyt-dta) [inline pyais/messages.py:967]: are these lat/long as well, and should use suitable to/from functions?

File: pyais/messages.py [hunk 2]
```diff
@@ -784,8 +787,8 @@ class MessageType17(Payload):
     repeat = bit_field(2, int, default=0)
     mmsi = bit_field(30, int, from_converter=from_mmsi, to_converter=to_mmsi)
     spare_1 = bit_field(2, int, default=0)
-    lon = bit_field(18, int, from_converter=from_course, to_converter=to_course, default=0)
-    lat = bit_field(17, int, from_converter=from_course, to_converter=to_course, default=0)
+    lon = bit_field(18, float, from_converter=from_course, to_converter=to_course, default=0)
```
Comment (jeremyt-dta) [inline pyais/messages.py:790]: should be from_lat_lon, to_lat_lon
Are there any test cases for these messages where data could be checked?

File: pyais/messages.py [hunk 3]
```diff
@@ -590,7 +593,7 @@ class MessageType5(Payload):
     day = bit_field(5, int, default=0)
     hour = bit_field(5, int, default=0)
     minute = bit_field(6, int, default=0)
-    draught = bit_field(8, int, from_converter=from_course, to_converter=to_course, default=0)
+    draught = bit_field(8, float, from_converter=from_course, to_converter=to_course, default=0)
```
Comment (jeremyt-dta) [inline pyais/messages.py:596]: should this have its own converter functions?

File: tests/test_decode.py [hunk 4]
```diff
@@ -344,10 +350,9 @@ def test_msg_type_17_b(self):
...
     def test_msg_type_18(self):
         msg = decode(b"!AIVDM,1,1,,A,B5NJ;PP005l4ot5Isbl03wsUkP06,0*76").asdict()
-        print(msg)
         assert msg['msg_type'] == 18
         assert msg['mmsi'] == "367430530"
-        assert msg['speed'] == 0
+        assert msg['speed'] == 0.0
         assert msg['accuracy'] == 0
         assert round(msg['lat'], 2) == 37.79
         assert round(msg['lon'], 2) == -122.27
```
Comment (jeremyt-dta) [inline tests/test_decode.py:358]: should the course (line below) be 0.0 now?

File: tests/test_encode.py [hunk 5]
```diff
@@ -508,6 +508,33 @@ def test_encode_type_19():
...
+        'lat': 37.785035,
+        'lon': -122.26732,
+        'mmsi': '367430530',
+        'msg22': 1,
+        'radio': 917510,
+        'raim': 0,
+        'regional': 0,
+        'repeat': 0,
+        'second': 55,
+        'speed': 67.85,
```
Comment (jeremyt-dta) [inline tests/test_encode.py:530]: speed should not have 2 decimal places, only one.

File: tests/test_decode.py [hunk 6]
```diff
@@ -597,7 +618,7 @@ def test_empty_channel(self):
         self.assertEqual(content["msg_type"], 18)
         self.assertEqual(content["repeat"], 0)
         self.assertEqual(content["mmsi"], "1000000000")
-        self.assertEqual(content["speed"], 1023)
+        self.assertEqual(content["speed"], 102.3)
```
Comment (jeremyt-dta) [inline tests/test_decode.py:621]: actually speed value of 1023 means not available, it does seem a bit strange seeing it as an actual float. But consistent with API use which doesn't explicitly deal with not provided values.

File: pyais/messages.py [hunk 7]
```diff
@@ -1299,8 +1302,8 @@ class MessageType27(Payload):
     accuracy = bit_field(1, int, default=0)
     raim = bit_field(1, bool, default=0)
     status = bit_field(4, int, default=0, from_converter=NavigationStatus, to_converter=NavigationStatus)
-    lon = bit_field(18, int, from_converter=from_lat_lon_600, to_converter=to_lat_lon_600, default=0)
-    lat = bit_field(17, int, from_converter=from_lat_lon_600, to_converter=to_lat_lon_600, default=0)
+    lon = bit_field(18, float, from_converter=from_lat_lon_600, to_converter=to_lat_lon_600, default=0)
+    lat = bit_field(17, float, from_converter=from_lat_lon_600, to_converter=to_lat_lon_600, default=0)
     speed = bit_field(6, int, default=0)
```
Comment (jeremyt-dta) [inline pyais/messages.py:1307]: speed / course here also missing float. Add a couple more test cases for the different types to confirm. I think:
Message type 1, 9, 18, 19, 27 have speed/course. I didn't check all lat/lon usage

**Review by jeremyt-dta — CHANGES_REQUESTED** [review]
Hi, a number of good changes for this problem, but still some missing, and potentially incorrect converter functions used. I think adding some additional tests would be useful, especially for values that are 'valid' vs 0 / not provided or out of range in the spec. They are of course also useful tests!
Is there any way to compare 'old' vs new decoder output on a sample of test data (vs test_decode function), or are the data types / outputs too different?

File: tests/test_decode.py (hunk 6 above)
Comment (Inrixia) [reply tests/test_decode.py:621]: Realistically 1023 should be returned as `None` if it represents not available.

File: pyais/messages.py (hunk 1 above)
Comment (M0r13n) [reply pyais/messages.py:967]: ne_lon, ne_lat, sw_lon, and sw_lat fields are in 0.1 minutes. [Source](https://gpsd.gitlab.io/gpsd/AIVDM.html#_type_22_channel_management)

File: pyais/messages.py (hunk 2 above)
Comment (M0r13n) [reply pyais/messages.py:790]: No, latitude and longitude are in units of a tenth of a minute. [Source](https://gpsd.gitlab.io/gpsd/AIVDM.html#_type_17_dgnss_broadcast_binary_message).

There is a testcase for it: https://github.com/M0r13n/pyais/blob/e833f46cbaa157aa022113c10a5630c002104459/tests/test_decode.py#L323

File: pyais/messages.py (hunk 3 above)
Comment (M0r13n) [reply pyais/messages.py:596]: Well, it would be a duplicate function that is identical to `to_course`. Instead I renamed `to_course` to `from_10th` and `to_10th`.

File: tests/test_encode.py (hunk 5 above)
Comment (M0r13n) [reply tests/test_encode.py:530]: That's part of the testcase actually. This testcase tests the encoding of a message. And a user might pass a float that has more than a single decimal place. But this should be handled by the lib - which it is.

If you decode `!AIVDO,1,1,,A,B5NJ;PP2aUl4ot5Isbl6GwsUkP06,0*35` you get `67.8`.

File: tests/test_decode.py (hunk 4 above)
Comment (M0r13n) [reply tests/test_decode.py:358]: In Python `0 == 0.0` is always `True`. So it doesn't even matter. The relevant test for ensuring that the values are floats comes a couple of lines below:

```py
assert isinstance(msg['speed'], float)
assert isinstance(msg['course'], float)
```

File: tests/test_decode.py (hunk 6 above)
Comment (M0r13n) [reply tests/test_decode.py:621]: I think that this is something that the application should decide. I would like to return the raw values as often as possible. Also both [tbsalling](http://ais.tbsalling.dk/) and [aggsoft](https://www.aggsoft.com/ais-decoder.htm), two decoders that I used heavily during my development, return `102.3` as a value. I think that it would be wise to be consistent here.

File: pyais/messages.py (hunk 7 above)
Comment (M0r13n) [reply pyais/messages.py:1307]: No, both `speed` and `course` are unsigned integers in this case. [Source](https://gpsd.gitlab.io/gpsd/AIVDM.html#_type_27_long_range_ais_broadcast_message).

There is a matching testcase:   [def test_msg_type_27(self):](https://github.com/M0r13n/pyais/blob/e833f46cbaa157aa022113c10a5630c002104459/tests/test_decode.py#L546)

Excerpt:

```
assert msg['lon'] == 137.023333
assert msg['lat'] == 4.84
assert msg['speed'] == 57
assert msg['course'] == 167
```

which is the same as [tbsalling](http://ais.tbsalling.dk/) returns.

**Comment by M0r13n** [comment]
@jeremyt-dta 

Hey, thank you for your suggestions. I looked at all of them and most were not quite correct. 

It is important to note how the respective field is to be interpreted **for each type**. Just because `lat` and `lon`  are given in Minutes/10000 most the time, this does not mean that these fields are **always** to be interpreted in this way. Good examples here are messages of type 17,22 or 27, where `lat` and `lon` are not encoded as a 10000th of a minute. Therefore it must be checked exactly for each message which conversion/encoding is to be used.

I went through your suggestions and checked them against https://gpsd.gitlab.io. It turned out that your suggestions were not correct.

> I think adding some additional tests would be useful, especially for values that are 'valid' vs 0 / not provided or out of range in the spec. 

As I stated somewhere above, I think that is would be wise to not interpret the values. This may be done by the application using `pyais`. This is because, many encoders/decoders seem to use the NMEA spec rather as a rough guideline than an actually spec. Therefore, I would like to keep the interpreting of invalid/valid data to a minimum. Also, this would obfuscate the type of values returned for each field even more difficult. Currently, each field can be either `None` (Optional) or have a single type (e.g. int, float, str). It would mess things up, if the lib would return additional values like enums or strings. Imagine the example above:

For a message of type 18, a value of `1023` for speed means that it is not available. How should one return this value to an user of the lib? Keep in mind that as a user I expect the lib to return a `float`. So if I get a string `invalid` oder `not available` my application is going to crash. Also returning `None` is not quite correct, because the field is not `None`. It is `1023`. I think that this is also why all of the popular online decoders return `102.3`.

**Comment by Inrixia** [comment]
> I think that this is also why all of the popular online decoders return `102.3`.

This makes sense and considering `102.3` is unique being well above any normal speed that ought to be reported and the likelihood of a error causing it is low I think its fine.

**Comment by jeremyt-dta** [comment]
ok great @M0r13n. The naming convention of converter functions had confused me - I went through all other lat/lon and speed/course values to check if they were the same, or potential copy/paste errors, rather than referring to the spec. In those cases the better name helps avoid potential confusion!
I was surprised that message 27 drops speed/course precision, but makes sense as the updates don't occur so much, and they are wanting to save bits. A comment to that effect (why values are stored differently) can help as well.
The reason for the interest is that we are using this code for AIS message decoding, so thanks a lot! We were trying to track down some bugs/differences in 1.0 and 2.0 behaviour for the class B messages in particular which should be resolved now.

**Comment by M0r13n** [comment]
@jeremyt-dta I appreciate that. :+1: 


The more eyes look at the code the better!
